### ISSUE TITLE
Issue with N64 Depth Compare enabled as default

### ISSUE THREAD

**Issue by HenWilOliPin** [gonetz/GLideN64#1970]
I ran into an issue recently where N64 Depth Compare being enabled by default caused m64p to not be able to startup any of my games, as detailed here:
https://github.com/m64p/mupen64plus-GLideN64/issues/115
loganmc10 suggested I file a report here, so figured I'd go ahead. Only way I got it to work again was by turning off N64 Depth Compare.

**Comment by loganmc10** [comment]
Can you post your OS and graphics card here as well?

**Comment by HenWilOliPin** [comment]
Sure, OS is Windows 10 and gfx card is Radeon RX 580, version 22.19.677.257

**Comment by loganmc10** [comment]
That is a somewhat old version of the GPU drivers is it not? Based on my reading, I assume updating the driver to something with a 24.x or 25.x version would fix the issue for you

**Comment by loganmc10** [comment]
To explain why I say that:

It appears that older versions of the RX 580 drivers declare support for the ```GL_INTEL_fragment_shader_ordering``` OpenGL extension, which when present, will allow N64 Depth Compare to work.

Later versions of the driver remove support for that extension. It's possible that it was included by accident, or perhaps it was so broken they decided to remove it from the driver.

By updating your GPU driver, support for that extension should go away, and N64 Depth Compare will disable itself automatically

**Comment by HenWilOliPin** [comment]
Interesting. Like I mentioned in the first thread, I just hit update in the device manager and let windows sort it out. Guess I'll try updating again or track down updated drivers manually, but for now I'll just keep it disabled.

**Comment by loganmc10** [comment]
I would just grab the latest drivers from AMD's site if I was you. If you're doing any kind of gaming it's best to have the latest drivers from the manufacturer

**Comment by gonetz** [comment]
> Radeon 17.x drivers support N64 style depth compare just fine¹, I used it before, the performance was good and there were never any glitches with it whatsoever. However, the latest 18.x, do not support N64 style depth compare.

This is bad news to know. I don't remember when I updated drivers for my Radeon. From my experience, its better to not update drivers if your card is more than 3 years old. So, I have no issues with N64 style depth compare, but if the problem exists with newer drivers, we can't have that option enabled by default. Users will not downgrade drivers just to solve that issue.

**Comment by loganmc10** [comment]
If they don't support it, it will just get disabled, the issue is when the GPU drivers have broken implementations of the needed extensions

**Comment by oddMLan** [comment]
>broken implementations

Well, like I said before N64 Depth Compare worked perfectly with the older driver versions, so I don't think that thier implementation for the needed extension was broken

**Comment by gonetz** [comment]
Two more issues with N64 depth compare found:
- 1080 two players node
- Indiana Jones

**Comment by Frank-74** [comment]
Got a new laptop i7 @3.7 Ghz, nVidia NVS 5200M, Windows 10 Pro x64.  It has OpenGL 4.6, but doesn't support N64 depth compare.
It also has integrated Intel HD 4000 Graphics, no depth compare support either. :(

**Comment by theboy181** [comment]
@gonetz 

1080 has issues with my HD Enhanced Hack.

It used to work back in March with no issues.

Please test with this GameShark Code on.

1080 Snowboarding (U)

Improve Draw Distance.
81109A5A 3F00
81109A5C 4612

Note: Broken and March update of GLideN64

This code increases the Draw Distance, and is part of the HD cheat I created for this game. Code works on console, and old versions of of GLideN64 from March. Looks like a possible regression.

Any thing you can do to find its cause would be helpful.

**Comment by loganmc10** [comment]
@theboy181 does the hack work with angrylion?

**Comment by theboy181** [comment]
Yes, and real HW. As I was explaining it looks like a regression from after the March release. :(

**Comment by oddMLan** [comment]
I tried one of the latest PCSX2 builds recently and they somehow managed to get PS2 Depth compare to a good speed, using the DirectX 11 backend. I remember it coming with a huge performance penalty using the OpenGL backend before, because similarly to older GLideN64 it wasn't caching the triangles. Maybe it's worth taking a look at how they managed to get it fast using DirectX? Although Idk how useful that would be for OpenGL GLideN64...

**Comment by oddMLan** [comment]
Found the relevant commits... apparently they use stencil buffer?

https://github.com/PCSX2/pcsx2/commit/4328d099c41ca10f02e79db1c702e8892ce06071
https://github.com/PCSX2/pcsx2/commit/ade00f8a7098f629a17e1f01f014733303cd74b2
https://github.com/PCSX2/pcsx2/commit/7928cff3546a913d5089515190a61f6abbc3372f
https://github.com/PCSX2/pcsx2/commit/c3a1cc3801153c38133f5226d7c2d08bc9400fcc
https://github.com/PCSX2/pcsx2/commit/7215618621df2def572ef3cccada3914bdb5e8e6

**Comment by theboy181** [comment]
Do they have any videos or screenshot comparisons?

**Comment by oddMLan** [comment]
There's still a performance penalty with the option enabled, maybe DX11+Depth Emulation it's just faster because AMD's OpenGL drivers suck.

**Comment by gonetz** [comment]
I fixed compatibility of N64 depth compare with Indiana and 1080.
I have no visible performance penalty when I enable that option: fullscreen + widescreen, works smooth.
I use Radeon drivers 17.7
Thus, the feature is ready to be set by default again. The only obstacle is bug in current Radeon drivers:
https://github.com/gonetz/GLideN64/issues/1970#issuecomment-449242717

**Comment by oddMLan** [comment]
Someone should nag AMD on their forums to re-enable the `GL_INTEL_fragment_shader_ordering` extension. I don't see a good reason why they would've disabled it in the first place, it wasn't buggy.

**Comment by fzurita** [comment]
@gonetz You can add driver version specific checks like done here:

https://github.com/gonetz/GLideN64/blob/f787483dbb8467d093ceccd8e460319bbb16d507/src/Graphics/OpenGLContext/opengl_GLInfo.cpp#L61-L76

**Comment by gonetz** [comment]
I can, but I don't like such things.
It is better to write good FAQ with setup instructions.

**Comment by oddMLan** [comment]
@fzurita That is not necessary, N64 Depth Compare will get disabled automatically if the driver won't support it. That would help only if there was an alternate codepath, like the Depth Compare without triangle caching, the way it was before, but it is incredibly taxing and probably not worth it.

**Comment by fzurita** [comment]
@oddMLan my comment was in regards disable N64 depth compare with drivers that say that they have support for the correct extensions but the extensions don't work correctly.

**Comment by loganmc10** [comment]
N64 Depth Compare currently requires 1 of 3 possible extensions

GL_INTEL_fragment_shader_ordering
shader_interlock
framebuffer_fetch

Im on mobile, I can't remember the full name of those last 2, but Google will tell you

**Comment by gonetz** [comment]
Mario Golf in-game depth compare issues with N64 depth compare enabled.

**Comment by gonetz** [comment]
Mario Golf uses depth buffer copy, see issue #312.
It is well emulated with standard FBO texture depth attachment, since it can be fetched in shaders as regular texture. Image texture requires special treatment, as it was done for BAR fog.

**Comment by gonetz** [comment]
I see no more issues with N64 depth compare except for Mario Golf.
Thus, it can be enabled as default again and disabled for Mario Golf in custom.ini

**Comment by gonetz** [comment]
@loganmc10 I started to test current version before release and found that many games have performance issues with N64 depth compare enabled. For example, Banjo Tooie (U) runs full-speed without it, but VI drops to below 40 with it. It is pity because that game runs more correct with that option.

**Comment by loganmc10** [comment]
Yeah I suppose even with the improvements it may still not be fast enough on every game for every PC.

I can't think of any further ways to optimize it, unless there is some opportunity to simplify the shader or solve the problem in a completely different manner.

**Comment by gonetz** [comment]
Yes, I understand it. My point is that N64 Depth Compare should not be enabled as default atm, even if all graphics glitches will be fixed.

**Comment by Craftyawesome** [comment]
Another issue on goldeneye results.
Depth compare on
![n64 depth compare on](https://user-images.githubusercontent.com/19976035/63620937-b3a77680-c5c0-11e9-9c3f-f7973d94e939.png)
Depth compare off
![n64 depth compare off](https://user-images.githubusercontent.com/19976035/63620945-b86c2a80-c5c0-11e9-8a8a-70ff7532fc0a.png)

**Comment by gonetz** [comment]
My system decided to upgrade graphics drivers. So. I have 19.x now.
17.x is the last known version of Radeon drivers, which supported necessary extensions to run N64 depth compare. Pity. I'm thinking to return the old implementation of N64 depth compare, which works slow but without extra extensions.

**Comment by fzurita** [comment]
Can both methods exist at the same time and the fast can be used if the right extensions exist?

**Comment by loganmc10** [comment]
@gonetz are you able to generate a list of the extensions your graphics card supports? Sometimes vendors have their own implementations of GL extensions, I could see if the card supports any alternatives that might work

It's been so long since I've looked at this, but it looks like it currently requires GL_EXT_shader_framebuffer_fetch and (GL_ARB_fragment_shader_interlock || GL_INTEL_fragment_shader_ordering)

**Comment by loganmc10** [comment]
Actually reading back through the history, I guess the issue is that newer versions of the AMD drivers remove support for GL_INTEL_fragment_shader_ordering

I wouldn't remove the current fast version, I guess you could probably add back support for the slow code path, but you'd probably have to create a new option where people need to specifically enable it, because there are games that have it enabled in the custom.ini file, and most people would rather have it disabled then use the slow code path

**Comment by oddMLan** [comment]
https://github.com/GPUOpen-Drivers/AMDVLK/issues/108#issuecomment-524159358
Looks like they had a reason to remove it. Fuck AMD :(

**Comment by Degerz** [comment]
@gonetz Actually, your old implementation of N64 depth compare might be faster on AMD HW since GL_fragment_shader_ordering or similar is a proven slow path on AMD HW ...

Also, I doubt GL_Intel_fragment_shader_ordering truly worked as intended on AMD HW prior to GCN5/Vega since I saw in a research paper (I can't remember where exactly) where the authors encountered ordering bugs when using that extension on AMD GPUs. If you used the equivalent feature in D3D11.3/12 where AMD properly implements the feature you'll get a real idea of just bad it is to use that feature on AMD HW ...

**Comment by gonetz** [comment]
> Can both methods exist at the same time and the fast can be used if the right extensions exist?

Yes. I already re-implemented it and I'm testing it now.

> I wouldn't remove the current fast version, I guess you could probably add back support for the slow code path

Yes, it will fall into slow code path when ext_fetch is false, image textures are supported but fragment ordering extensions are not.

> you'd probably have to create a new option where people need to specifically enable it, because there are games that have it enabled in the custom.ini file, and most people would rather have it disabled than use the slow code path

It is a good point, thanks.

> Actually, your old implementation of N64 depth compare might be faster on AMD HW since GL_fragment_shader_ordering or similar is a proven slow path on AMD HW

As I remember, optimized  N64 depth compare worked pretty fast on my hardware with older drivers.
I will not re-install the drivers to check it though. I just need this feature to run on my current hardware with current drivers, no matter slow or fast. 

> the authors encountered ordering bugs when using that extension on AMD GPUs.

We have this issue: #2066. It looks as ordering bugs, but it was first reported for NVidia and then confirmed for AMD.

**Comment by Degerz** [comment]
> As I remember, optimized N64 depth compare worked pretty fast on my hardware with older drivers.
I will not re-install the drivers to check it though. I just need this feature to run on my current hardware with current drivers, no matter slow or fast.

In all likelihood it was a buggy implementation if you were using AMD's OpenGL drivers. Only GCN5/Vega truly supports fragment shader ordering or interlocks as people now know them because AMD just recently added an SOPP opcode to support the feature ... 

> We have this issue: #2066. It looks as ordering bugs, but it was first reported for NVidia and then confirmed for AMD.

Also it sounds like you're missing a memory barrier after every draw call ? Did you make sure to insert a memory barrier after every draw call ? This is absolutely required because interlocks don't automatically synchronize access to texture data at draw call granularity.

**Comment by gonetz** [comment]
> Also it sounds like you're missing a memory barrier after every draw call ? Did you make sure to insert a memory barrier after every draw call ? This is absolutely required because interlocks don't automatically synchronize access to texture data at draw call granularity.

Ops! Probably not. Do you mean glMemoryBarrier or memoryBarrier for GLSL?

**Comment by Degerz** [comment]
> Ops! Probably not. Do you mean glMemoryBarrier or memoryBarrier for GLSL?

That depends on your use case. 

I found [this](https://www.khronos.org/opengl/wiki/GLAPI/glMemoryBarrier) on the OpenGL wiki maintained by Khronos.

> The following guidelines may be helpful in choosing when to use coherent memory accesses and when to use barriers.
> - Data that are read-only or constant may be accessed without using coherent variables or calling glMemoryBarrier. Updates to the read-only data via API calls such as BufferSubData will invalidate shader caches implicitly as required. 
> - Data that are shared between shader invocations at a fine granularity (e.g., written by one invocation, consumed by another invocation) should use coherent variables to read and write the shared data. 
> - Data written by one shader invocation and consumed by other shader invocations launched as a result of its execution ("dependent invocations") should use coherent variables in the producing shader invocation and call memoryBarrier() after the last write. The consuming shader invocation should also use coherent variables. 
> - Data written to image variables in one rendering pass and read by the shader in a later pass need not use coherent variables or memoryBarrier(). Calling glMemoryBarrier with the SHADER_IMAGE_ACCESS_BARRIER_BIT set in barriers​ between passes is necessary. 
> - Data written by the shader in one rendering pass and read by another mechanism (e.g., vertex or index buffer pulling) in a later pass need not use coherent variables or memoryBarrier(). Calling glMemoryBarrier with the appropriate bits set in barriers​ between passes is necessary.

**Comment by gonetz** [comment]
Truly speaking, I'm not quite understand that explanation by Khronos.
Our scenario is quite simple. The plugin uses image texture to keep depth buffer value. Each fragment shader has "depth_compare" function, which reads from and writes to that texture. The function does not use memoryBarrier. The old _slow_ method of synchronization was to draw polygons one-by-one and to call glMemoryBarrier(GL_SHADER_IMAGE_ACCESS_BARRIER_BIT) between draw calls. Logan implemented current _fast_ method using various techniques and extensions. New method insures the synchronization when several polygons rendered by a single draw command, and thus works way faster.  Old method was removed. As I see, the current code does not use glMemoryBarrier and memoryBarrier at all.

**Comment by Degerz** [comment]
It sounds like you want to revert back to your old method or just use glMemoryBarrier on your existing solution ... 

TBH, interlocks sounds like one of those ideas that are good on paper but then turns out to be sh*t because of slow hardware implementation like geometry shaders or it doesn't cover cases like this in the end. Once Intel enters the discrete GPU market and if they _ever disrecommend_ using interlocks _altogether_ in the future, I would probably at that point pretend that the feature never exists and never use it again because the idea is already deeply unpopular with AMD and Nvidia. It'd be a crazy turn to see Intel who originally championed (only one left as well to care) this feature to be standardized only for them advise against using it in future HW but that aside since it's getting off-topic ... 

I guess were all out of options until desktop IHVs implement "tile memory" just like mobile GPUs to make the case of ordered access between draw calls faster as well ? 

If still want to maximize performance without any advanced features, you still can try implementing [per-pixel linked lists](https://hgpu.org/?p=2066) and store the blend parameters (need this to store blend states), depth, color, next id (serves as a pointer to the previous fragment), primitive order id (need this for sorting) and then sort the fragments by primitive order but that's going to be more complex than any of your previous solutions.

**Comment by Jj0YzL5nvJ** [comment]
Another problematic case #2074

**Comment by roguen1** [comment]
> Two more issues with N64 depth compare found:
> 
>     * 1080 two players node
> 
>     * Indiana Jones

Sorry for bringing up this old topic, but I was wondering if anyone else had experienced similar glitches with n64depthcompare, and it seems they have. For example, I’ve seen the black background in 007 GoldenEye after dying, but in 1080º or Indiana Jones I didn’t notice anything strange, was it at the beginning of the game? And from what I understand, based on what you said, it seems like these issues are specific to AMD graphics cards? Because yeah, I have an RX 6700XT, but would it still be safe to keep n64depthcompare enabled globally even if there are glitches in occasional games? From what I understand, it might do more good than harm (I think)
